src/main/java/org/tugraz/sysds/runtime/data/DenseBlockLBool.java

```diff
@@ -0,0 +1,366 @@
+		}
+		if (neededCompleteBlocks == _blocks.length) {
+			// Check if we can store enough rows in the last (most likely smaller) block
+			// and we don't need another row.
+			return neededLastBlockSize == 0 && possibleRowsPerBlock == lastBlockSize;
+		}
+		return false;
+	}
+
+	@Override
+	public void reset(int rlen, int[] odims, double v) {
+		boolean bv = v != 0;
+		if(!isReusable(rlen, odims)) {
+		    int newBlockSize = Integer.MAX_VALUE / odims[0];
+		    int restBlockSize = rlen % newBlockSize;
+		    int newNumBlocks = (rlen / newBlockSize) + (restBlockSize == 0 ? 0 : 1);
+		    _fullBlockSize = newBlockSize;
+			if (restBlockSize == 0) {
+				_lastBlockSize = newBlockSize;
+				_blocks = new BitSet[newNumBlocks];
+				for (int i = 0; i < newNumBlocks; i++) {
+					_blocks[i] = new BitSet(newBlockSize * odims[0] + 1);
+					_blocks[i].set(newBlockSize * odims[0] + 1);
+				}
+			} else {
+				_lastBlockSize = restBlockSize;
+				_blocks = new BitSet[newNumBlocks];
+				for (int i = 0; i < newNumBlocks - 1; i++) {
+					_blocks[i] = new BitSet(newBlockSize * odims[0] + 1);
+					_blocks[i].set(newBlockSize * odims[0] + 1);
+				}
+				_blocks[newNumBlocks - 1] = new BitSet(restBlockSize * odims[0] + 1);
+				_blocks[newNumBlocks - 1].set(restBlockSize * odims[0] + 1);
+			}
+			if( bv ) {
+				for (int i = 0; i < newNumBlocks; i++) {
+				    _blocks[i].set(0, _blocks[i].length() - 1);
+				}
+			}
+		}
+		else {
+			// Memory is enough, overwrite
+			int newBlockSize = Integer.MAX_VALUE - 1 / odims[0];
+			int restBlockSize = rlen % newBlockSize;
+			int newNumBlocks = (rlen / newBlockSize) + (restBlockSize == 0 ? 0 : 1);
+			_fullBlockSize = newBlockSize;
+			if (restBlockSize == 0) {
+				_lastBlockSize = newBlockSize;
+				for (int i = 0; i < newNumBlocks; i++) {
+					_blocks[i].set(0, newBlockSize * odims[0], bv);
+					_blocks[i].set(newBlockSize * odims[0] + 1);
+				}
+			} else {
+				_lastBlockSize = restBlockSize;
+				for (int i = 0; i < newNumBlocks - 1; i++) {
+					_blocks[i].set(0, newBlockSize * odims[0], bv);
+					_blocks[i].set(newBlockSize * odims[0] + 1);
+				}
+				_blocks[newNumBlocks - 1].set(0, restBlockSize * odims[0], bv);
+				_blocks[newNumBlocks - 1].set(restBlockSize * odims[0] + 1);
+			}
+		}
+		_rlen = rlen;
+		_odims = odims;
+	}
+
+	@Override
+	public int numBlocks() {
+		return _blocks.length;
+	}
+
+	@Override
+	public int blockSize() {
+	    return (_blocks.length == 1) ? _lastBlockSize : _fullBlockSize;
+	}
+
+	@Override
+	public int blockSize(int bix) {
+		return (bix == _blocks.length - 1) ? _lastBlockSize : _fullBlockSize;
+	}
+
+	@Override
+	public long capacity() {
+		return (_blocks!=null) ? (long)(_blocks.length - 1) * _blocks[0].size() + _blocks[_blocks.length - 1].size() : -1;
+	}
+
+	@Override
+	public long countNonZeros() {
+		long nnz = 0;
+		for (int i = 0; i < _blocks.length - 1; i++) {
+			nnz += UtilFunctions.computeNnz(_blocks[i], 0, blockSize() * _odims[0]);
```

Comment:
if we ensure that all bits past the logical size of a block are zero, we could simply call Bitset.cardinality.